rcl/src/rcl/node.c

```diff
@@ -115,7 +128,25 @@ const char * rcl_get_secure_root(const char * node_name, const rcl_allocator_t *
   if (!ros_secure_root_size) {
     return NULL;  // environment variable was empty
   }
-  char * node_secure_root = rcutils_join_path(ros_secure_root_env, node_name, *allocator);
+  char * node_secure_root = NULL;
+  // TODO(ros2team): This make assumption on the value and length of the root namespace.
+  // This should likely come from another (rcl/rmw?) function for reuse.
+  // If the namespace is the root namespace ("/"), the secure root is just the node name.
+  if (strlen(node_namespace) == 1) {
+    node_secure_root = rcutils_join_path(ros_secure_root_env, node_name, *allocator);
+  } else {
+    char * node_fqn = NULL;
+    char * node_root_path = NULL;
+    // Combine node namespace with node name
+    // TODO(ros2team): remove the hard-coded value of the root namespace.
+    node_fqn = rcutils_format_string(*allocator, "%s%s%s", node_namespace, "/", node_name);
+    // Get native path, ignore the leading forward slash.
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead.
+    node_root_path = rcutils_to_native_path(node_fqn + 1, *allocator);
+    node_secure_root = rcutils_join_path(ros_secure_root_env, node_root_path, *allocator);
+    allocator->deallocate(node_fqn, allocator->state);
+    allocator->deallocate(node_root_path, allocator->state);
+  }
   if (!rcutils_is_directory(node_secure_root)) {
```

Comment:
I know you didn't add this, but there is a problem with `NULL` here.  Assume that `rcutils_join_path` at line 136 returns NULL.  In that case, the `node_secure_root` will be NULL, which will be passed into `rcutils_is_directory`, which will be then passed directly to `stat`, which will most likely core dump.  A similar problem exists in the sequence in 138-148.  I think we need to check for `NULL` right after creating `node_secure_root`, and return `NULL` if it is (before the `rcutils_is_directory` check).